tenpy/algorithms/mps_compress.py

```diff
@@ -0,0 +1,345 @@
+
+
+
+
+def make_WII(t, A, B, C, D):
+	r""" WII approx to exp(t H) from sys (A, B, C, D)
+
+        Parameters
+        ----------
+        t : float
+            time intervall for the propagator
+        A : ndarray
+        B : ndarray
+        C : ndarray
+        D : ndarray
+		
+	"""
+	### Algorithm
+	#
+	# In the paper, we have two formal parameter "phi_{r/c}" which satisfies phi_r^2 = phi_c^2 = 0
+	# To implement this, we temporarily extend the virtual Hilbert space with two hard-core bosons "br, bl"
+	# The components of Eqn (11) can be computed for each index of the virtual row / column independently
+	# The matrix exponential is done in the hard-core extended Hilbert space
+	
+	tB = t/np.sqrt(np.abs(t)) #spread time step across B, C
+	tC = np.sqrt(np.abs(t))
+	d = D.shape[0]
+	
+	#The virtual size of W is  (1+Nr, 1+Nc)
+	Nr = A.shape[0]
+	Nc = A.shape[1]
+	W = np.zeros( (1 + Nr, 1 + Nc, d, d), dtype = np.result_type(D, t))
+
+	Id_ = np.array([[1, 0], [0, 1]]) #2x2 operators in a hard-core boson space
+	b = np.array([[0, 0], [1, 0]])
+
+	Id = np.kron(Id_, Id_) #4x4 operators in the 2x hard core boson space
+	Br = np.kron(b, Id_)
+	Bc = np.kron(Id_, b)
+	Brc = np.kron(b, b)
+	for r in xrange(Nr):    #double loop over row / column of A
+		for c in xrange(Nc):
+			#Select relevent part of virtual space and extend by hardcore bosons
+			h = np.kron(Brc, A[r, c, :, :]) + np.kron(Br, tB*B[r, :, :]) + np.kron(Bc, tC*C[c, :, :]) + t*np.kron(Id, D)
+			w = sp.linalg.expm(h) #Exponentiate in the extended Hilbert space
+			w = w.reshape((2, 2, d, 2, 2, d))
+			w = w[:, :, :, 0, 0, :]
+			W[1+r, 1+c, :, :] = w[1, 1]  #This part now extracts relevant parts according to Eqn 11
+			if c==0:
+				W[1+r, 0] = w[1, 0]
+			if r==0:
+				W[0, 1+c] = w[0, 1]
+				if c==0:
+					W[0, 0] = w[0, 0]
+		if Nc==0: #technically only need one boson
+			h = np.kron(Br, tB*B[r, :, :]) + t*np.kron(Id, D)
+			w = sp.linalg.expm(h)
+			w = w.reshape((2, 2, d, 2, 2, d))
+			w = w[:, :, :, 0, 0, :]
+			W[1+r, 0] = w[1, 0]
+			if r==0:
+				W[0, 0] = w[0, 0]
+
+	if Nr == 0:
+		for c in xrange(Nc):
+			h = np.kron(Bc, tC*C[c, :, :]) + t*np.kron(Id, D)
+			w = sp.linalg.expm(h)
+			w = w.reshape((2, 2, d, 2, 2, d))
+			w = w[:, :, :, 0, 0, :]
+			W[0, 1+c] = w[0, 1]
+			if c==0:
+				W[0, 0] = w[0, 0]
+		if Nc==0:
+			W =sp.linalg.expm(t*D)
+
+	return W
+
+
+def mps_compress(psi, trunc_par):
+    r"""Takes an MPS and compresses it. In Place.
+
+    Parameters
+    ----------
+    psi : :class:`tenpy.networks.mps.MPS`
+        MPS to be compressed.
+        The singular values are usually ignored and need not be set to sensible values, except for bc='infinite', where psi.S[0] is needed as an initial guess.
+    trunc_par : dict
+        See :func:`~tenpy.algorithms.truncation.truncate`
+    """
+    bc=psi.bc
+    L=psi.L
+    if bc=='finite':
+        # Do QR starting from the left
+        B=psi.get_B(0,form=None)

```

Comment:
Probably better to use the form explicitly...

Reply:
The problem was, that the form was not correct anyway, as the contraction of MPO tensors and MPS tensors had no canonical form. The only thing I know and use is, that the full contraction in the given form results in the correct state without inserting any S values.

Reply:
Yes. The problem I see, however, is that one might also use this function in other contexts than after `apply_mpo`, where the singular values are/might be correct. Then we should take proper care of the form to get the correct results....

Reply:
Then I could take the 'B' form here except for the first one in the finite case where I need the theta to get all S values. In apply_mpo I would then set exactly those forms, although it's wrong. Then this part works for a correct MPS.